src/services/accessibility/html_id_generator.ts

```diff
@@ -23,3 +24,22 @@ export function htmlIdGenerator(idPrefix: string = '') {
     return `${prefix}${suffix ? staticUuid : uuidv1()}${suffix}`;
   };
 }
+
+/**
+ * Generates an ID within a static ref object that remains static
+ * until the component using it is unmounted. This prevents IDs from
+ * being re-randomized on every component update.
+ */
+export const useGeneratedHtmlId = ({
```

Comment:
Since we already have a `htmlIdGenerator` service, it would be great to keep the hook similarly named so that it's easier to find similar instances.
```suggestion
export const useHtmlIdGenerator = ({
```

Reply:
Ah, so I started a thread on this earlier here (https://github.com/elastic/eui/pull/5133#discussion_r701229253) on why `useHtmlIdGenerator` naming felt wrong (which I now realize I shouldn't have pre-emptively collapsed 😅)

Basically my reasoning is that the actual usage and output of the use hook vs. the original function is significantly different, so I changed the name deliberately to make that distinction clearer. Here's why they're different:

- `useGeneratedHtmlId()` at the end of the day outputs a **memoized string**. I think `generatedHtmlId` makes that clearer.
- `useHtmlIdGenerator()` outputs a **generator function** that can be called multiple different times to generate multiple IDs that will use the same prefix UUID. The hook does not and **will always generate completely different IDs**.

As an example, a few EUI components currently do something like this:

```js
import React from 'react';

import { htmlIdGenerator } from '../../services';

const generateId = htmlIdGenerator();

export class SomeComponent {
  tooltipId = generateId('tooltip');
  panelId = generateId('panel');
  // etc.
}
```

[EuiSelectable](https://github.com/elastic/eui/blob/005f9e3afd9e2e5a6cc178b3fae053a00bbf75b0/src/components/selectable/selectable.tsx#L156) is the best example I found from a quick scan that reuses the initial UUID generator to create more suffixed child IDs. Those components likely don't want to use the current use hook as is, and need to retain usage of the original generator functionality.

On the other hand, there are plenty of other components that are simple and straightforward and just need 1 single randomly generated ID, no prefix/suffix needed. For those cases, `useGeneratedHtmlId` is perfect and they can be converted to using that. But I want to be super deliberate about that choice, as the prefix/suffix/curried functionality of `htmlIdGenerator` isn't obvious at first glance, and I think distinguishing that in the naming will help prevent further confusion down the road.

Reply:
There seems to then be more nuances to this hook vs the original service that may be good to detail in the docs since I just interpreted the new hook as literally just the hook version of the same service.

Reply:
Yeah, it's definitely not! To be fair though I don't think the original generator is _super_ thoroughly documented either and some of the nuances are lost 😅 For example, if you pass in the same suffix to a generator twice separately, it will output the same ID each time, which I thought was kind of a weird choice (why not just save the generated ID to a var if you want the same ID twice?)

I'll add another commit more clearly documenting the following:

- the original generator and more of its nuances
- the specific differences between the generator and the use hook
- more examples of the use hook params (primarily the custom ID prop, once we figure out naming)

Reply:
👋 I added a new section in https://github.com/elastic/eui/pull/5133/commits/e5049faefe40fbc21acfeabb2700f5b43f506dc1 with an example of creating multiple IDs from the same generator, and added a new paragraph to the hook section explaining the differences and also what it's best used for. Let me know if you think that's sufficient!

<img width="1189" alt="" src="https://user-images.githubusercontent.com/549407/132607975-c89fbfda-b1bd-45ec-805d-ea0a71b7c111.png">

Reply:
🙌 Nice!